lib/hanami/view/context.rb

```diff
@@ -13,67 +14,12 @@ class View
     #
     # @api public
     class Context
-      include Dry::Equalizer(:_options)
+      include Dry::Effects.Reader(:render_env)
+      include Dry::Effects.Reader(:scope)
       include DecoratedAttributes
 
-      attr_reader :_render_env, :_options
-
-      # Returns a new instance of Context
-      #
-      # In subclasses, you should include an `**options` parameter and pass _all
-      # arguments_ to `super`. This allows Context to make copies of itself
-      # while preserving your dependencies.
-      #
-      # @example
-      #   class MyContext < Hanami::View::Context
-      #     # Injected dependency
-      #     attr_reader :assets
-      #
-      #     def initialize(assets:, **options)
-      #       @assets = assets
-      #       super
-      #     end
-      #   end
-      #
-      # @api public
-      def initialize(render_env: nil, **options)
-        @_render_env = render_env
-        @_options = options
-      end
-
-      # @api private
-      def for_render_env(render_env)
-        return self if render_env == _render_env
-
-        self.class.new(**_options.merge(render_env: render_env))
-      end
-
-      # Returns a copy of the Context with new options merged in.
-      #
-      # This may be useful to supply values for dependencies that are _optional_
-      # when initializing your custom Context subclass.
-      #
-      # @example
-      #   class MyContext < Hanami::View::Context
-      #     # Injected dependencies (request is optional)
-      #     attr_reader :assets, :request
-      #
-      #     def initialize(assets:, request: nil, **options)
-      #       @assets = assets
-      #       @request = reuqest
-      #       super
-      #     end
-      #   end
-      #
-      #   my_context = MyContext.new(assets: assets)
-      #   my_context_with_request = my_context.with(request: request)
-      #
-      # @api public
-      def with(**new_options)
-        self.class.new(
-          render_env: _render_env,
-          **_options.merge(new_options)
-        )
+      def locals
+        scope._locals
```

Comment:
I think, for performance reasons, it makes sense to memoize `scope`. Otherwise, you'll resolve it via an effect every time you access `scope` (or `locals`). For application code, it won't be a problem in 99% of cases, but I would be more cautious in a library. Of course, this is only applicable if `Context` is a class for one-per-render objects

Reply:
Unfortunately, while context is indeed a one-instance-per-render object, the `scope` here will change multiple times over the course of that render: for every template or partial, in fact.

See the related code over in `RenderEnvironment` to see this in action:

```ruby
      # TODO:
      # - Give this keyword arguments, it's getting out of control
      # - Rename `scope` to `scope_name` to make it clear it's not the fully built scope
      def template(name, scope, locals, &block)
        template_env = chdir(name)
        scope = self.scope(scope, locals)

        with_render_env(template_env) {
          with_scope(scope) {
            renderer.template(name, scope, &block)
          }
        }
      end

      # TODO: Update this to match how `#template` works and see what the flow-on effects are
      def partial(name, scope, &block)
        with_scope(scope) {
          renderer.partial(name, scope, &block)
        }
      end
```

So unfortunately I don't think we can get away with memoizing here.

Your comment about performance makes me think I should probably look at a building and running a few benchmarks test before merging this code in. Does that seem reasonable to you? Is there anything in particular I should be looking out for, performance-wise?